### DISCUSSION TITLE
Coffee editor ID vs XMI id

### DISCUSSION THREAD

**Discussion by LokanandaHJ** [eclipse-glsp/glsp#951] [General]
What is the difference between ID and XMI id that is used in superbrewer.coffee file

previously coffee editor was using XMI ID in superbrewer.coffee file
![MicrosoftTeams-image](https://user-images.githubusercontent.com/97349849/224022617-5969d78a-28a5-4468-880b-04fd2b459ab6.png)

now it is changed to ID 
![MicrosoftTeams-image (1)](https://user-images.githubusercontent.com/97349849/224022693-50ea88fa-3671-4456-b129-e4f601829b66.png)

we are using XMI ID in our implementation

please explain the significance of this ID

will it affect the Error markers in the notation file

**Comment by ndoschek** [comment]
Hi @LokanandaHJ,

With an earlier update of the coffee-editor (https://github.com/eclipse-emfcloud/coffee-editor/pull/457), the coffee model was updated to use ID attributes for `EObjects`.

Previously, some custom code was needed (e.g. `CoffeeResourceImpl`, `CoffeeResource`) to generate unique element ids for the `xmi:id` field.
Now this was replaced by the Ecore ID attribute, which is a built-in EMF feature.
![image](https://user-images.githubusercontent.com/47414588/224266624-cc01cc9b-b0fe-47d7-9ff0-e7b53753dc3d.png)

As the ID attribute is a generic feature, it is also easier to determine model elements via the ID and vice versa, because EMF utilities can be used (e.g. `EcoreUtil.getID(eObject)`).

This ID attribute is now used instead of the `xmi:id` to generate the graphical model elements in the `WorkflowGModelFactory`.
This will not affect the markers of the diagram, as the markers are also based on the same ID attribute and hence the mapping will still work.

HTH and kind regards,
Nina

**Reply by Jaritha** [reply]
Hi @ndoschek,

We are using **XMI IDs** only. We have observed that if **ID** is not available then EMF cloud validation framework is not mapping the markers.

We are trying to validate the GLSP Nodes and mark the Errors by using Validation framework as shown below. The identifier is always returning "/-1" because its always checking for ID attribute. 

![image](https://user-images.githubusercontent.com/45856518/224998091-a9a87f27-be9f-4a1c-a323-11fac3f8fe10.png)


**Observation :** We are able to see the markers for the first time as shown in the below diagram. On refresh the framework is unable to find the ID and is returning "/-1". 

![image](https://user-images.githubusercontent.com/45856518/225001224-3b5d6ba5-a6ca-4476-a1b7-dd569596d16b.png)


**Requesting your suggestions on how to validate GLSP nodes based on XMI ID.**


Thank you!

**Reply by ndoschek** [reply]
Hi @Jaritha,
If it works for the first time the validation is run and cannot find the identifier on reload, it sounds like you do not use consistent graphical element IDs.
Could you check if you set the actual model element id (i.e. in your case the `xmi:id`) as graphical element ID in the `WorkflowGModelFactory`?

E.g.: `build.id(toId(task));`
`toId(..)` should retrieve the `xmi:id` of your semantic model element, such that the graphical elements get the same id each time you load the diagram.

Best,
Nina
